src/transformers/cache_utils.py

```diff
@@ -468,15 +602,62 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
         # this part of code otherwise fails when used to verify attn_weight shape in some models
         return self._seen_tokens if layer_idx == 0 else self._seen_tokens - 1
 
-    def _quantize(self, tensor, axis):
+    def _quantize(self, tensor, axis, cache_type):
         """Quantizes a key/value using a defined quantization method."""
         raise NotImplementedError("Make sure to implement `_quantize` in a subclass.")
 
-    def _dequantize(self, q_tensor):
+    def _dequantize(self, q_tensor, cache_type):
         """Dequantizes back the tensor that was quantized by `self._quantize()`"""
         raise NotImplementedError("Make sure to implement `_dequantize` in a subclass.")
 
 
+class CompressedTensorsCache(QuantizedCache):
+    def __init__(self, cache_config: CacheConfig) -> None:
+        super().__init__(cache_config)
+
+        self.kv_cache_scheme = cache_config.kv_cache_scheme
+        self.scales_per_layer = cache_config.scales_per_layer
+        self.zero_points_per_layer = cache_config.zero_points_per_layer
+        self.quant_dtype = self._establish_quant_dtype(
+            num_bits=self.kv_cache_scheme.num_bits, type_=self.kv_cache_scheme.type
+        )
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # temporarily save the current layer_idx as a class attribute,
+        # so it can be reused in _quantize/_dequantize methods
+        self._layer_idx = layer_idx
+        return super().update(key_states, value_states, layer_idx, cache_kwargs)
+
+    @staticmethod
+    def _establish_quant_dtype(num_bits: int, type_: str) -> torch.dtype:
+        if num_bits == 8 and type_ == "int":
```

Comment:
We should potentially add more supported quantization types, to be discussed @bfineran